src/python/pants/vcs/git.py

```diff
@@ -106,6 +116,58 @@ def changed_files(
         # git will report changed files relative to the worktree: re-relativize to relative_to
         return {self._fix_git_relative_path(f, relative_to) for f in files}
 
+    def changed_files_lines(
+        self,
+        paths: Iterable[str],
+        /,
+        *,
+        from_commit: str | None = None,
+        relative_to: PurePath | str | None = None,
+        include_untracked: bool = False,
+    ) -> dict[str, tuple[Hunk, ...]]:
+        relative_to = PurePath(relative_to) if relative_to is not None else self.worktree
+
+        result = self._git_diff(
+            "--unified=0",
+            "HEAD",
+            "--",
+            *[str(relative_to / path) for path in paths],
+        )
+
+        if from_commit:
+            diff = self._git_diff(
+                "--unified=0",
+                from_commit + "...HEAD",
+                "--",
+                *[str(relative_to / path) for path in paths],
+            )
+            result.update(diff)
+
+        if include_untracked:
+            # There is no git diff flag to include untracked files, so we get
+            # the list of untracked files and manually create the diff by
+            # comparing each file to an empty /dev/null.
+            untracked_files = self._git(
```

Comment:
We could maybe pull this out to a helper function shared with changed_files(), but it's not that important. We can do it later.